File: backend/core/monitoring.py
"""Sentry monitoring bootstrap.

The SDK is intentionally dormant unless a DSN is configured. This lets the same
source run in local/dev/test without emitting telemetry, while production can
turn monitoring on through environment variables only.
"""
from __future__ import annotations

import os
from typing import Any, Mapping

_SENSITIVE_KEYS = (
    "authorization",
    "cookie",
    "password",
    "token",
    "secret",
    "api_key",
    "apikey",
    "refresh",
    "stripe",
    "docusign",
    "storage_access_key",
    "storage_secret",
)


def sentry_configured(env: Mapping[str, str] | None = None) -> bool:
    e = os.environ if env is None else env
    return bool((e.get("SENTRY_DSN") or "").strip())


def _redact_mapping(value: Any) -> Any:
    if isinstance(value, dict):
        clean = {}
        for key, inner in value.items():
            lowered = str(key).lower()
            if any(marker in lowered for marker in _SENSITIVE_KEYS):
                clean[key] = "[Filtered]"
            else:
                clean[key] = _redact_mapping(inner)
        return clean
    if isinstance(value, list):
        return [_redact_mapping(item) for item in value]
    return value


def _before_send(event: dict[str, Any], hint: dict[str, Any]) -> dict[str, Any] | None:
    request = event.get("request") or {}
    headers = request.get("headers") or {}
    if isinstance(headers, dict):
        for key in list(headers):
            if any(marker in str(key).lower() for marker in _SENSITIVE_KEYS):
                headers[key] = "[Filtered]"
    event["extra"] = _redact_mapping(event.get("extra") or {})
    event["contexts"] = _redact_mapping(event.get("contexts") or {})
    return event


def _traces_sampler(sampling_context: dict[str, Any]) -> float:
    tx = sampling_context.get("transaction_context") or {}
    name = str(tx.get("name") or "")
    if any(path in name for path in ("/api/health", "/api/health/live", "/api/health/ready")):
        return 0.0
    # Keep pilot traffic visible without turning every request into telemetry.
    return float(os.environ.get("SENTRY_TRACES_SAMPLE_RATE", "0.2"))


def init_sentry() -> bool:
    dsn = (os.environ.get("SENTRY_DSN") or "").strip()
    if not dsn:
        return False

    import sentry_sdk

    sentry_sdk.init(
        dsn=dsn,
        environment=os.environ.get("SENTRY_ENVIRONMENT", os.environ.get("APP_ENV", "production")),
        release=os.environ.get("SENTRY_RELEASE") or os.environ.get("RENDER_GIT_COMMIT"),
        send_default_pii=False,
        attach_stacktrace=True,
        traces_sampler=_traces_sampler,
        before_send=_before_send,
        enable_logs=True,
    )
    sentry_sdk.set_tag("service", "equinesync-api")
    return True

File: frontend/src/monitoring.js
import * as Sentry from "@sentry/react";

const dsn = process.env.REACT_APP_SENTRY_DSN;
const environment = process.env.REACT_APP_SENTRY_ENVIRONMENT || process.env.NODE_ENV || "production";
const release = process.env.REACT_APP_SENTRY_RELEASE;
const apiUrl = process.env.REACT_APP_BACKEND_URL;

export const sentryEnabled = Boolean(dsn);

const sensitiveKeyPattern = /(authorization|cookie|password|token|secret|api[_-]?key|stripe|docusign|storage)/i;

const scrub = (value) => {
  if (Array.isArray(value)) return value.map(scrub);
  if (!value || typeof value !== "object") return value;
  return Object.fromEntries(
    Object.entries(value).map(([key, inner]) => [
      key,
      sensitiveKeyPattern.test(key) ? "[Filtered]" : scrub(inner),
    ]),
  );
};

const sampleTrace = (context) => {
  const name = context?.name || "";
  if (name.includes("/health")) return 0;
  if (name.includes("/billing") || name.includes("/subscription") || name.includes("/checkout")) return 1.0;
  return Number(process.env.REACT_APP_SENTRY_TRACES_SAMPLE_RATE || 0.2);
};

if (sentryEnabled) {
  Sentry.init({
    dsn,
    environment,
    release,
    integrations: [
      Sentry.browserTracingIntegration(),
      Sentry.replayIntegration({
        maskAllText: true,
        blockAllMedia: true,
      }),
    ],
    tracesSampler: sampleTrace,
    tracePropagationTargets: [
      "localhost",
      /^https:\/\/api\.equine-sync\.com\/api/,
      ...(apiUrl ? [apiUrl] : []),
    ],
    replaysSessionSampleRate: Number(process.env.REACT_APP_SENTRY_REPLAY_SESSION_SAMPLE_RATE || 0),
    replaysOnErrorSampleRate: Number(process.env.REACT_APP_SENTRY_REPLAY_ON_ERROR_SAMPLE_RATE || 1.0),
    beforeSend(event) {
      event.extra = scrub(event.extra || {});
      event.contexts = scrub(event.contexts || {});
      return event;
    },
  });
  Sentry.setTag("service", "equinesync-web");
}

export const reactRootErrorHandler = sentryEnabled ? Sentry.reactErrorHandler() : undefined;
